File: sdk/typescript/src/webhook.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

export const DEFAULT_MAX_AGE_SECS = 300;
export const CLOCK_SKEW_SECS = 60;

export function signWebhook(secret: string, timestamp: string, body: Buffer | Uint8Array): string {
  const mac = createHmac('sha256', secret);
  mac.update(timestamp);
  mac.update('.');
  mac.update(body);
  return `sha256=${mac.digest('hex')}`;
}

export function verifySignature(
  secret: string,
  signature: string,
  timestamp: string,
  body: Buffer | Uint8Array,
  options?: { now?: number; maxAgeSecs?: number },
): boolean {
  if (!secret || !signature || !timestamp) {
    return false;
  }
  if (!signature.startsWith('sha256=')) {
    return false;
  }
  const ts = Number.parseInt(timestamp, 10);
  if (!Number.isFinite(ts)) {
    return false;
  }
  const now = options?.now ?? Date.now() / 1000;
  const maxAge = options?.maxAgeSecs ?? DEFAULT_MAX_AGE_SECS;
  const age = now - ts;
  if (age > maxAge || ts - now > CLOCK_SKEW_SECS) {
    return false;
  }
  const expected = signWebhook(secret, timestamp, body);
  try {
    const got = Buffer.from(signature.slice(7), 'hex');
    const want = Buffer.from(expected.slice(7), 'hex');
    if (got.length !== want.length) {
      return false;
    }
    return timingSafeEqual(got, want);
  } catch {
    return false;
  }
}

File: sdk/python/pcmi/webhook.py
"""Webhook signature verification for PCMI HTTP deliveries."""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Union

DEFAULT_MAX_AGE_SECS = 300
CLOCK_SKEW_SECS = 60


def verify_signature(
    secret: str,
    signature: str,
    timestamp: str,
    body: bytes,
    *,
    now: float | None = None,
    max_age_secs: int = DEFAULT_MAX_AGE_SECS,
) -> bool:
    """Verify X-PCMI-Signature for a webhook POST body.

    Signature format: sha256={hex(HMAC-SHA256(secret, timestamp + "." + body))}
    """
    if not secret or not signature or not timestamp:
        return False
    if not signature.startswith("sha256="):
        return False
    try:
        ts = int(timestamp)
    except ValueError:
        return False
    now_ts = time.time() if now is None else now
    age = now_ts - ts
    if age > max_age_secs or ts - now_ts > CLOCK_SKEW_SECS:
        return False
    expected = _sign(secret, timestamp, body)
    got_hex = signature[7:]
    try:
        got = bytes.fromhex(got_hex)
        want = bytes.fromhex(expected[7:])
    except ValueError:
        return False
    return hmac.compare_digest(got, want)


def _sign(secret: str, timestamp: str, body: bytes) -> str:
    msg = timestamp.encode() + b"." + body
    digest = hmac.new(secret.encode(), msg, hashlib.sha256).hexdigest()
    return f"sha256={digest}"
